tests/e2e/statefulsets.go

```diff
@@ -985,6 +1046,589 @@ var _ = ginkgo.Describe("statefulset", func() {
+					gomega.Expect(err).NotTo(gomega.HaveOccurred(), "Disk is not attached to the node")
+					gomega.Expect(isDiskAttached).To(gomega.BeTrue(), "Disk is not attached")
+					ginkgo.By("After scale up, verify the attached volumes match those in CNS Cache")
+					err = verifyVolumeMetadataInCNS(&e2eVSphere, pv.Spec.CSI.VolumeHandle,
+						volumespec.PersistentVolumeClaim.ClaimName, pv.ObjectMeta.Name, sspod.Name)
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+					if stretchedSVC {
+						verifyAnnotationsAndNodeAffinityInSVC(allowedTopologyHAMap, pod, nodeList, pv)
+					}
+
+				}
+			}
+		}
+		replicas = 0
+		ginkgo.By(fmt.Sprintf("Scaling down statefulsets to number of Replica: %v", replicas))
+		_, scaledownErr = fss.Scale(ctx, client, statefulset, replicas)
+		gomega.Expect(scaledownErr).NotTo(gomega.HaveOccurred())
+		fss.WaitForStatusReplicas(ctx, client, statefulset, replicas)
+		ssPodsAfterScaleDown = fss.GetPodList(ctx, client, statefulset)
+		gomega.Expect(len(ssPodsAfterScaleDown.Items) == int(replicas)).To(gomega.BeTrue(),
+			"Number of Pods in the statefulset should match with number of replicas")
+	})
+
+	/**
+	Statefulset statefullset nodeAffinity
+	1. create zonal storage class
+	2. create statefull set  with replica 3 using zonal sc along with node affinity details
+	3. scale down statefull set
+	4. scale up statefull set
+	5. Verify node affinity on each PV
+	6. Verify statefull pods are coming up on appropriate nodes
+	7. clean up the data
+	*/
+
+	ginkgo.It("[stretched-svc] statefulset-nodeAffinity", ginkgo.Label(p0, wcp, core), func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		ginkgo.By("Creating StorageClass for Statefulset")
+
+		storageClassName = zonalPolicy
+
+		ginkgo.By("Creating service")
+		service := CreateService(namespace, client)
+		defer func() {
+			deleteService(namespace, client, service)
+		}()
+
+		statefulset := GetStatefulSetFromManifest(namespace)
+		ginkgo.By("Creating statefulset")
+		statefulset.Spec.VolumeClaimTemplates[len(statefulset.Spec.VolumeClaimTemplates)-1].
+			Spec.StorageClassName = &storageClassName
+
+		scParameters[svStorageClassName] = zonalPolicy
+		storageclass, err := client.StorageV1().StorageClasses().Get(ctx, zonalPolicy, metav1.GetOptions{})
+		if !apierrors.IsNotFound(err) {
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+		storageClassName = storageclass.Name
+
+		topologyHaMap := GetAndExpectStringEnvVar(topologyHaMap)
+
+		_, categories = createTopologyMapLevel5(topologyHaMap)
+		allowedTopologies := createAllowedTopolgies(topologyHaMap)
+		allowedTopologyHAMap = createAllowedTopologiesMap(allowedTopologies)
+		allowedTopologies = getTopologySelector(allowedTopologyHAMap, categories,
+			stretchedSVCTopologyLevels)
+		allowedTopologyHAMap = createAllowedTopologiesMap(allowedTopologies)
+
+		storageClassName = storageclass.Name
+		statefulset.Spec.VolumeClaimTemplates[len(statefulset.Spec.VolumeClaimTemplates)-1].
```

Comment:
For everywhere statefulset creation please use the utils mentioned above. Feel free to make changes to these utils if it is not satisfying your requirement.

Lets avoid writing repeated code everywhere